enhancements/machine-api/machine-deletion-hooks.md

```diff
@@ -0,0 +1,424 @@
+
+## Proposal
+
+- Utilize annotations to implement lifecycle hooks.
+- Each lifecycle point can have 0 or more hooks.
+- Hooks do not enforce ordering.
+- Hooks found during machine reconciliation effectively pause reconciliation
+until all hooks for that lifecycle point are removed from a machine's annotations.
+
+### User Stories
+
+#### Story 1
+(pre-terminate) As an operator, I would like to have the ability to perform
+different actions between the time a Machine is marked deleted in the API and
+the time the Machine is deleted from the infrastructure provider.
+
+For example, when replacing a Control Plane Machine, ensure a new Control
+Plane Machine has been successfully created and joined to the cluster before
+removing the instance of the deleted Machine. This might be useful in case
+there are disruptions during replacement and we need the disk of the existing
+instance to perform some disaster recovery operation.  This will also prevent
+prolonged periods of having one fewer Control Plane host in the event the
+replacement instance does not come up in a timely manner.
+
+#### Story 2
+(pre-terminate) As an operator, I would to have the ability to perform some
+action between the time that a Machine has been drained, but before it is
+removed from the infrastructure provider.
+
+For example, when storage is attached to the Machine due to some pod placement,
+an operator may want to check that the storage is detached before allowing the
+removal of the machine from the infrastructure provider.
+
+Alternatively, once Nodes are drained, an additional delay may need to be added
+to allow the Node's log exporter daemon to synchronise all logs to the
+centralised logging system. A pre-terminate hook could allow a log operator to
+ensure the main workloads are removed and no longer adding to the log backlog
+before waiting for the log exporter to catch up on the synchronisation process,
+ensuring all application logs are captured.
+
+#### Story 3
+(pre-drain) As an operator, I want the ability to utilize my own draining
+controller instead of the logic built into the machine controller. This will
+allow me better flexibility and control over the lifecycle of workloads on each
+node.
+
+For example this would allow me to prioritise moving particular mission critical
+applications first to ensure that service interruptions are minimised in cases
+where cluster capacity is limited. Ordering is not supported today in drain
+libraries and as such, this would require a custom drain provider.
+
+### Implementation Details/Notes/Constraints
+
+For each defined lifecycle point, one or more hooks may be applied as an annotation
+to the machine object. These annotations will pause reconciliation of a Machine object
+until all hooks are resolved for that lifecycle point.
+The hooks should be managed by a Hook Implementing Controller or other external
+application, or manually created and removed by an administrator.
+
+#### Lifecycle Points
```

Comment:
the different lifecycle points are a potentially good reason to avoid finalizers (thank you for the alternatives section).  But if you have a structured shape, why is it better to use an annotation instead of creating a spec field indicating what you want to have happen at these points?

Reply:
To give more context on hypershift a capi machine or any other capi resource is considered an internal impl detail. NodePool is the consume facing interface and this hook concept does not exist.
Also the only known real use case is control plane and etcd which does not apply to hypershift since the control plane run as pods.